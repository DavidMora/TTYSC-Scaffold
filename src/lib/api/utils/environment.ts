/**
 * Environment utility functions for API routes
 */

/**
 * Determines the current environment based on NODE_ENV and BACKEND_BASE_URL
 * @returns Environment type: 'dev' | 'stg' | 'prd'
 */
export function getEnvironment(): 'dev' | 'stg' | 'prd' {
  const env = process.env.NODE_ENV;
  if (env === 'production') {
    // Check for specific environment indicators
    if (process.env.BACKEND_BASE_URL?.includes('stg')) return 'stg';
    if (process.env.BACKEND_BASE_URL?.includes('prd')) return 'prd';
    return 'prd'; // Default for production
  }
  return 'dev';
}

/**
 * Gets the current environment configuration
 */
export function getEnvironmentConfig() {
  return {
    type: getEnvironment(),
    isProduction: process.env.NODE_ENV === 'production',
    isDevelopment: process.env.NODE_ENV !== 'production',
    backendBaseUrl: process.env.BACKEND_BASE_URL || '',
    feedbackWorkflowName: process.env.FEEDBACK_WORKFLOW_NAME || 'ttysc',
  };
}
