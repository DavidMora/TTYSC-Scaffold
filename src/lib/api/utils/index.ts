/**
 * Shared API utilities for eliminating duplication across API routes
 */

// Environment utilities
export * from './environment';

// Response utilities
export * from './response';

// Authentication utilities
export * from './auth';
