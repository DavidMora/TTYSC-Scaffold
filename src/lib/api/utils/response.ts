/**
 * Shared response utility functions for API routes
 */

/**
 * Creates a standardized error response
 * @param error - The error object or message
 * @param status - HTTP status code (default: 500)
 * @returns Response object with JSON error
 */
export function createErrorResponse(error: unknown, status = 500): Response {
  let message: string;
  if (typeof error === 'string') message = error;
  else if (error instanceof Error) message = error.message;
  else message = 'Internal server error';

  return new Response(JSON.stringify({ error: message }), {
    status,
    headers: { 'Content-Type': 'application/json' },
  });
}

/**
 * Creates a standardized success response with JSON data
 * @param data - The response data
 * @param status - HTTP status code (default: 200)
 * @returns Response object with JSON data
 */
export function createJsonResponse<T = unknown>(
  data: T,
  status = 200
): Response {
  return new Response(JSON.stringify(data), {
    status,
    headers: { 'Content-Type': 'application/json' },
  });
}

/**
 * Creates a response from upstream backend data
 * @param upstream - The upstream response data
 * @returns Response object with upstream data and merged headers
 */
export function createUpstreamResponse<T = unknown>(upstream: {
  data: T;
  status: number;
  headers?: Record<string, string>;
}): Response {
  // Clone and merge upstream headers, then enforce JSON content type
  const headers = new Headers(upstream.headers || {});
  headers.set('Content-Type', 'application/json');

  return new Response(JSON.stringify(upstream.data), {
    status: upstream.status,
    headers,
  });
}

/**
 * Common response utilities for API routes
 */
export const apiResponse = {
  error: createErrorResponse,
  json: createJsonResponse,
  upstream: createUpstreamResponse,
};
