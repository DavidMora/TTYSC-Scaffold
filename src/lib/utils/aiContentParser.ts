import { ParsedContentItem } from '../types/chatContent';

const showTableRegex = /\[SHOW_TABLE\]/gi;
// Regex updated to capture optional UUID: [SHOW_CHART] or [SHOW_CHART:uuid]
const showChartRegex =
  /\[SHOW_CHART(?::([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}))?\]/gi;

function parseCodeBlock(
  text: string,
  startIndex: number
): ParsedContentItem | null {
  const endIndex = text.indexOf('```', startIndex + 3);
  if (endIndex === -1) return null;

  const blockContent = text.substring(startIndex + 3, endIndex);
  const lines = blockContent.split('\n');

  let language = '';
  let codeContent = blockContent;

  if (lines.length > 0) {
    const firstLine = lines[0].trim();
    if (firstLine && /^\w+$/.test(firstLine)) {
      language = firstLine;
      codeContent = lines.slice(1).join('\n');
    }
  }

  return {
    type: 'code',
    language: language,
    content: codeContent,
    index: startIndex,
    matchLength: endIndex - startIndex + 3,
  };
}

function parseCodeBlocks(text: string): ParsedContentItem[] {
  const results: ParsedContentItem[] = [];
  let currentIndex = 0;

  while (currentIndex < text.length) {
    const startIndex = text.indexOf('```', currentIndex);
    if (startIndex === -1) break;

    const codeBlock = parseCodeBlock(text, startIndex);
    if (codeBlock) {
      results.push(codeBlock);
      currentIndex = codeBlock.index + (codeBlock.matchLength || 0);
    } else {
      currentIndex = startIndex + 3;
    }
  }

  return results;
}

function parseRegexMatches(text: string): ParsedContentItem[] {
  const results: ParsedContentItem[] = [];

  showTableRegex.lastIndex = 0;
  let match;
  while ((match = showTableRegex.exec(text)) !== null) {
    if (match.index === showTableRegex.lastIndex) {
      showTableRegex.lastIndex++;
    }

    results.push({
      type: 'table',
      index: match.index,
      matchLength: match[0].length,
    });
  }

  showChartRegex.lastIndex = 0;
  while ((match = showChartRegex.exec(text)) !== null) {
    if (match.index === showChartRegex.lastIndex) {
      showChartRegex.lastIndex++;
    }

    results.push({
      type: 'chart',
      index: match.index,
      matchLength: match[0].length,
      chartId: match[1],
    });
  }

  return results;
}

export function parseContent(text: string): ParsedContentItem[] {
  const MAX_CONTENT_LENGTH = 1000000;
  if (text.length > MAX_CONTENT_LENGTH) {
    console.warn('Content too long, truncating to prevent performance issues');
    text = text.substring(0, MAX_CONTENT_LENGTH);
  }

  const codeBlocks = parseCodeBlocks(text);
  const regexMatches = parseRegexMatches(text);

  return [...codeBlocks, ...regexMatches].sort((a, b) => a.index - b.index);
}
