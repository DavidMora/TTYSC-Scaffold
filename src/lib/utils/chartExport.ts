import {
  ChartDimension,
  ChartMeasure,
  MultiDataPoint,
  SingleDataPoint,
} from "@/lib/types/charts";

export function sanitizeFilename(name: string): string {
  return name.replace(/[^a-z0-9\-_.]+/gi, "_").slice(0, 120);
}

export function triggerFileDownload(
  data: string | Blob,
  filename: string,
  mime?: string
) {
  const blob =
    typeof data === "string"
      ? new Blob([data], { type: mime || "application/octet-stream" })
      : data;
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

export function downloadChartAsPng(
  container: HTMLElement | null,
  title?: string
) {
  const svg = findFirstSvg(container);
  if (!svg) return;
  serializeSvgToPng(svg, title || "chart");
}

export function getCurrentSlice<T>(
  data: T[],
  viewWindow: { start: number; end: number }
): T[] {
  const len = data.length;
  const from = Math.floor(viewWindow.start * len);
  const to = Math.ceil(viewWindow.end * len);
  const safeFrom = Math.max(0, Math.min(from, len - 1));
  const safeTo = Math.max(safeFrom + 1, Math.min(to, len));
  return data.slice(safeFrom, safeTo);
}

export function buildCsv(
  dataset: Array<SingleDataPoint | MultiDataPoint>,
  dimensions: ChartDimension[],
  measures: ChartMeasure[]
): string {
  const dimCols = dimensions.map((d) => d.accessor);
  const measureCols = measures.map((m) => m.accessor);
  const headers = [...dimCols, ...measures.map((m) => m.label)];
  const cols = [...dimCols, ...measureCols];
  const rows = dataset.map((row) =>
    cols
      .map((col) => toCsvString((row as Record<string, unknown>)[col]))
      .map((value) => escapeCsvField(value))
      .join(",")
  );
  return [headers.join(","), ...rows].join("\n");
}

function toCsvString(value: unknown): string {
  if (value == null) return "";
  if (typeof value === "object") {
    try {
      return JSON.stringify(value);
    } catch {
      return "";
    }
  }
  if (
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  ) {
    return String(value);
  }
  return "";
}

function escapeCsvField(raw: string): string {
  const value = raw.replace(/\r?\n/g, "\\n");
  const needsQuote = /[",\n]/.test(value);
  const doubledQuotes = value.replace(/"/g, '""');
  return needsQuote ? `"${doubledQuotes}"` : doubledQuotes;
}

function serializeSvgToPng(svg: SVGSVGElement, filenameBase: string) {
  const rect = svg.getBoundingClientRect();
  const width = Math.ceil(
    rect.width || Number(svg.getAttribute("width")) || 800
  );
  const height = Math.ceil(
    rect.height || Number(svg.getAttribute("height")) || 400
  );

  const cloned = svg.cloneNode(true) as SVGSVGElement;
  cloned.setAttribute("xmlns", "http://www.w3.org/2000/svg");
  cloned.setAttribute("width", String(width));
  cloned.setAttribute("height", String(height));
  const svgData = new XMLSerializer().serializeToString(cloned);
  const svgBlob = new Blob([svgData], { type: "image/svg+xml;charset=utf-8" });
  const url = URL.createObjectURL(svgBlob);

  const img = new Image();
  img.onload = () => {
    const canvas = document.createElement("canvas");
    if (canvas && typeof canvas === "object") {
      canvas.width = width;
      canvas.height = height;
    }
    const hasGetContext = canvas && typeof canvas.getContext === "function";
    const hasToBlob = canvas && typeof canvas.toBlob === "function";

    if (!hasGetContext || !hasToBlob) {
      triggerFileDownload(
        svgBlob,
        sanitizeFilename(filenameBase) + ".svg",
        "image/svg+xml;charset=utf-8"
      );
      URL.revokeObjectURL(url);
      return;
    }

    const ctx = canvas.getContext("2d");
    if (!ctx) {
      URL.revokeObjectURL(url);
      return;
    }
    ctx.fillStyle = "#ffffff";
    ctx.fillRect(0, 0, width, height);
    ctx.drawImage(img, 0, 0);
    canvas.toBlob((blob: Blob | null) => {
      if (blob) {
        triggerFileDownload(
          blob,
          sanitizeFilename(filenameBase) + ".png",
          "image/png"
        );
      }
      URL.revokeObjectURL(url);
    }, "image/png");
  };
  img.onerror = () => URL.revokeObjectURL(url);
  img.src = url;
}

function findFirstSvg(root: HTMLElement | null): SVGSVGElement | null {
  return (root?.querySelector("svg") as SVGSVGElement) || null;
}
