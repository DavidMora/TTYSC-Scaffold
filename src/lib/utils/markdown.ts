import { JSDOM } from 'jsdom';
import createDOMPurify from 'dompurify';
import { marked } from 'marked';
import { SANITIZER_CONFIG } from '@/lib/utils/sanitizer.config';

// Configure Marked for GFM (no server-side syntax highlighting)
marked.setOptions({ gfm: true });

// Singleton JSDOM/DOMPurify for performance
type SharedWindow = Window & typeof globalThis;
let sharedWindow: SharedWindow | null = null;
let DOMPurifyInstance: ReturnType<typeof createDOMPurify> | null = null;

type SanitizerAttributeHookEvent = {
  attrName?: string;
  keepAttr?: boolean;
  forceRemove?: boolean;
};

// Branded type to indicate content has been sanitized.
// This helps ensure only sanitized HTML flows into rendering sinks.
export type SafeHtml = string & { __brand: 'SafeHtml' };

function getDOMPurify(): ReturnType<typeof createDOMPurify> {
  if (!sharedWindow) {
    const dom = new JSDOM(
      '<!DOCTYPE html><html><head></head><body></body></html>'
    );
    sharedWindow = dom.window as unknown as SharedWindow;
  }
  if (!DOMPurifyInstance) {
    const purifier = createDOMPurify(sharedWindow as unknown as SharedWindow);
    // Configure once
    purifier.setConfig({
      ALLOWED_TAGS: SANITIZER_CONFIG.ALLOWED_TAGS,
      ALLOWED_ATTR: SANITIZER_CONFIG.ALLOWED_ATTR,
      FORBID_ATTR: SANITIZER_CONFIG.FORBID_ATTR,
      ALLOWED_URI_REGEXP: SANITIZER_CONFIG.ALLOWED_URI_REGEXP,
      ALLOW_DATA_ATTR: false,
      SAFE_FOR_TEMPLATES: true,
      FORBID_TAGS: SANITIZER_CONFIG.FORBID_TAGS,
    } as unknown as Record<string, unknown>);

    // Forbid any event handlers
    purifier.addHook(
      'uponSanitizeAttribute',
      (_node: Element, data: SanitizerAttributeHookEvent) => {
        if (data.attrName && /^on/i.test(data.attrName)) {
          data.keepAttr = false;
          data.forceRemove = true;
        }
      }
    );

    // Block SVG data URLs and other dangerous data URIs
    purifier.addHook(
      'uponSanitizeAttribute',
      (_node: Element, data: SanitizerAttributeHookEvent) => {
        if (data.attrName === 'src' || data.attrName === 'href') {
          const attrValue = _node.getAttribute(data.attrName);
          if (attrValue && /^data:image\/svg\+xml/i.test(attrValue)) {
            data.keepAttr = false;
            data.forceRemove = true;
          }
        }
      }
    );

    // Enforce anchor target/rel
    purifier.addHook('afterSanitizeAttributes', (node: Element) => {
      if (node.tagName === 'A') {
        const href = node.getAttribute('href') ?? '';
        // Only treat absolute external links as external
        const isExternal = /^https?:\/\//i.test(href) || href.startsWith('//');
        if (isExternal) {
          node.setAttribute('target', '_blank');
          const existingRel = node.getAttribute('rel') ?? '';
          const rel = new Set(
            existingRel
              .split(/\s+/)
              .filter(Boolean)
              .concat(['noopener', 'noreferrer', 'nofollow'])
          );
          node.setAttribute('rel', Array.from(rel).join(' '));
        } else if (node.getAttribute('target') === '_blank') {
          node.removeAttribute('target');
        }
      }
    });

    DOMPurifyInstance = purifier;
  }
  return DOMPurifyInstance;
}

export async function renderMarkdownToSafeHtml(
  markdown: string
): Promise<SafeHtml> {
  const rawHtml = await marked.parse(markdown ?? '');
  const purifier = getDOMPurify();

  const sanitized = purifier.sanitize(rawHtml, { RETURN_TRUSTED_TYPE: false });
  return sanitized as SafeHtml;
}
