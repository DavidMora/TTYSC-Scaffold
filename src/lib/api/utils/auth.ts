/**
 * Authentication utility functions for API routes
 */
import { getServerSession } from 'next-auth/next';
import { authOptions } from '@/lib/auth/auth-options';
import { createErrorResponse } from './response';

/**
 * Validates that a user is authenticated
 * @returns The session object if authenticated, null otherwise
 */
export async function validateSession() {
  const session = await getServerSession(authOptions);
  return session;
}

/**
 * Validates that a user is authenticated and has an email
 * @returns The user's email if authenticated, null otherwise
 */
export async function getAuthenticatedUserEmail(): Promise<string | null> {
  const session = await validateSession();
  return session?.user?.email || null;
}

/**
 * Creates an authentication error response
 * @returns Response with 401 Unauthorized
 */
export function createAuthErrorResponse(): Response {
  return createErrorResponse('Authentication required', 401);
}

/**
 * Validates authentication and returns user email or error response
 * @returns Object with either userEmail or errorResponse
 */
export async function requireAuthentication(): Promise<
  { userEmail: string } | { errorResponse: Response }
> {
  const userEmail = await getAuthenticatedUserEmail();
  if (!userEmail) {
    return { errorResponse: createAuthErrorResponse() };
  }
  return { userEmail };
}
