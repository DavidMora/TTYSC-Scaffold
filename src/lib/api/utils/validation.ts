import { NextRequest } from 'next/server';
import { apiResponse } from './response';

/**
 * Safely parses JSON from a NextRequest
 * @param req The NextRequest to parse
 * @returns Parsed JSON object or error response
 */
export async function parseJsonBody(
  req: NextRequest
): Promise<Record<string, unknown> | Response> {
  try {
    const result = await req.json();
    if (
      result === null ||
      typeof result !== 'object' ||
      Array.isArray(result)
    ) {
      return apiResponse.error(
        'Invalid request body: expected JSON object',
        400
      );
    }
    return result as Record<string, unknown>;
  } catch {
    return apiResponse.error('Invalid JSON in request body', 400);
  }
}

/**
 * Validates that a value is a non-null object (not array)
 * @param body The value to validate
 * @returns true if valid object, false otherwise
 */
export function isValidObjectBody(
  body: unknown
): body is Record<string, unknown> {
  return body !== null && typeof body === 'object' && !Array.isArray(body);
}

/**
 * Validates that required fields exist in an object
 * @param body The object to check
 * @param fields Array of required field names
 * @returns Missing fields array or null if all fields present
 */
export function validateRequiredFields(
  body: Record<string, unknown>,
  fields: string[]
): string[] | null {
  const missing = fields.filter((field) => !body[field]);
  return missing.length > 0 ? missing : null;
}

/**
 * Validates request body structure and returns standardized error if invalid
 * @param body The request body to validate
 * @returns Error response if invalid, null if valid
 */
export function validateObjectBody(body: unknown): Response | null {
  if (!isValidObjectBody(body)) {
    return apiResponse.error('Invalid request body: expected JSON object', 400);
  }
  return null;
}

/**
 * Validates required fields and returns standardized error if missing
 * @param body The object containing the fields
 * @param requiredFields Array of required field names
 * @returns Error response if missing fields, null if all present
 */
export function validateRequiredFieldsError(
  body: Record<string, unknown>,
  requiredFields: string[]
): Response | null {
  const missing = validateRequiredFields(body, requiredFields);
  if (missing) {
    return apiResponse.error(
      `Missing required fields: ${missing.join(', ')}`,
      400
    );
  }
  return null;
}

/**
 * Validates that specified fields are all strings
 * @param body The object containing the fields
 * @param fields Array of field names that should be strings
 * @returns Error response if any field is not a string, null if all are strings
 */
export function validateStringFields(
  body: Record<string, unknown>,
  fields: string[]
): Response | null {
  const invalidFields = fields.filter(
    (field) => typeof body[field] !== 'string'
  );

  if (invalidFields.length > 0) {
    return apiResponse.error(
      `Invalid field types: ${invalidFields.join(', ')} must be strings`,
      400
    );
  }

  return null;
}
