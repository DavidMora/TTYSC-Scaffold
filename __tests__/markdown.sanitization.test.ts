import { renderMarkdownToSafeHtml } from "@/lib/utils/markdown";

describe("Markdown sanitization", () => {
  it("removes <script> tags", async () => {
    const md = "Hello<script>alert(1)</script>World";
    const html = await renderMarkdownToSafeHtml(md);
    expect(html).not.toContain("<script>");
    expect(html).toContain("Hello");
    expect(html).toContain("World");
  });

  it("removes on* attributes", async () => {
    const md = '<img src="data:image/png;base64,i" onerror="alert(1)" />';
    const html = await renderMarkdownToSafeHtml(md);
    expect(html).not.toMatch(/on\w+=/i);
  });

  it("blocks javascript: URLs", async () => {
    const md = "[x](javascript:alert(1))";
    const html = await renderMarkdownToSafeHtml(md);
    // The anchor should lose href or be removed; ensure javascript: does not appear
    expect(html).not.toMatch(/javascript:/i);
  });

  it("adds target and rel to anchor tags", async () => {
    const md = "[example](https://example.com)";
    const html = await renderMarkdownToSafeHtml(md);
    expect(html).toMatch(/<a[^>]*target="_blank"/);
    expect(html).toMatch(/<a[^>]*rel="[^"]*noopener[^"]*noreferrer[^"]*nofollow[^"]*"/);
  });

  it("snapshot: tables and code blocks", async () => {
    const md = `
| h1 | h2 |
|----|----|
| a  | b  |

\`\`\`ts
const x: number = 1;
console.log(x);
\`\`\`
`;
    const html = await renderMarkdownToSafeHtml(md);
    expect(html).toMatchSnapshot();
  });

  it("blocks known tricky payloads", async () => {
    const payloads: string[] = [
      // outerHTML fragments with event handlers
      '<div onclick="alert(1)">Click</div>',
      '<img src="x" onerror="alert(1)">',
      // javascript: and mixed case variants
      '<a href="javascript:alert(1)">link</a>',
      '<a href="JaVaScRiPt:alert(1)">link</a>',
      // vbscript: (legacy)
      '<a href="vbscript:msgbox(1)">vb</a>',
      // malformed tags trying to smuggle script
      "<scr<script>ipt>alert(1)</scr</script>ipt>",
      // data:text payloads should not pass through
      "[x](data:text/html;base64,PHNjcmlwdD5hbGVydCgxKTwvc2NyaXB0Pg==)",
      // svg payloads (disallowed in URI allowlist)
      "![x](data:image/svg+xml;base64,PHN2ZyBvbmxvYWQ9YWxlcnQoMSk+PC9zdmc+)",
    ];

    for (const md of payloads) {
      const html = await renderMarkdownToSafeHtml(md);
      expect(html).not.toMatch(/<script/i);
      expect(html).not.toMatch(/\son\w+\s*=\s*/i);
      expect(html).not.toMatch(/javascript:/i);
      expect(html).not.toMatch(/vbscript:/i);
      expect(html).not.toMatch(/data:text\/html/i);
      expect(html).not.toMatch(/data:image\/svg\+xml/i);
    }
  });
});
