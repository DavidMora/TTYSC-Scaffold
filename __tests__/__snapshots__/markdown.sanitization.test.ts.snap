// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Markdown sanitization snapshot: tables and code blocks 1`] = `
"
<table>
<thead>
<tr>
<th>h1</th>
<th>h2</th>
</tr>
</thead>
<tbody><tr>
<td>a</td>
<td>b</td>
</tr>
</tbody></table>
<pre><code class="language-ts">const x: number = 1;
console.log(x);</code></pre>
"
`;
